pkg/highlight/highlighter.go

```diff
@@ -73,135 +73,180 @@ func findIndex(regex *regexp.Regexp, skip *regexp.Regexp, str []byte) []int {
-		if _, ok := highlights[0]; !ok {
-			highlights[0] = curRegion.group
+func (h *Highlighter) highlightRange(fullHighlights []Group, start int, end int, group Group) {
+	if start <= end && end <= len(fullHighlights) {
+		for i := start; i < end; i++ {
+			fullHighlights[i] = group
 		}
 	}
+}
 
-	var nestedRegion *region
-	nestedLoc := []int{lineLen, 0}
-	searchNesting := true
-	endLoc := findIndex(curRegion.end, curRegion.skip, line)
-	if endLoc != nil {
-		if start == endLoc[0] {
-			searchNesting = false
-		} else {
-			nestedLoc = endLoc
-		}
+func (h *Highlighter) highlightPatterns(fullHighlights []Group, start int, lineNum int, line []byte, curRegion *region) {
+	lineLen := util.CharacterCount(line)
+	if lineLen == 0 {
+		return
 	}
-	if searchNesting {
-		for _, r := range curRegion.rules.regions {
-			loc := findIndex(r.start, r.skip, line)
-			if loc != nil {
-				if loc[0] < nestedLoc[0] {
-					nestedLoc = loc
-					nestedRegion = r
-				}
+
+	var patterns []*pattern
+	if curRegion == nil {
+		patterns = h.Def.rules.patterns
+	} else {
+		patterns = curRegion.rules.patterns
+	}
+
+	for _, p := range patterns {
+		if curRegion == nil || curRegion.group == curRegion.limitGroup || p.group == curRegion.limitGroup {
+			matches := findAllIndex(p.regex, nil, line)
+			for _, m := range matches {
+				h.highlightRange(fullHighlights, start+m[0], start+m[1], p.group)
 			}
 		}
 	}
-	if nestedRegion != nil && nestedLoc[0] != lineLen {
-		highlights[start+nestedLoc[0]] = nestedRegion.limitGroup
-		slice := util.SliceEnd(line, nestedLoc[1])
-		h.highlightEmptyRegion(highlights, start+nestedLoc[1], lineNum, slice)
-		h.highlightRegion(highlights, start+nestedLoc[1], lineNum, slice, nestedRegion)
-		return highlights
-	}
+}
 
-	fullHighlights := make([]Group, lineLen)
-	for i := 0; i < len(fullHighlights); i++ {
-		fullHighlights[i] = curRegion.group
+func (h *Highlighter) highlightRegions(fullHighlights []Group, start int, lineNum int, line []byte, curRegion *region, regions []*region, nestedRegion bool) {
+	lineLen := util.CharacterCount(line)
+	if lineLen == 0 {
+		return
 	}
 
-	if searchNesting {
-		for _, p := range curRegion.rules.patterns {
-			if curRegion.group == curRegion.limitGroup || p.group == curRegion.limitGroup {
-				matches := findAllIndex(p.regex, line)
-				for _, m := range matches {
-					if endLoc == nil || m[0] < endLoc[0] {
-						for i := m[0]; i < m[1]; i++ {
-							fullHighlights[i] = p.group
+	h.highlightPatterns(fullHighlights, start, lineNum, line, curRegion)
+
+	lastStart := 0
+	regionStart := 0
+	regionEnd := 0
+	for _, r := range regions {
+		if !nestedRegion && curRegion != nil && (curRegion.group != r.group || curRegion.start != r.start) {
+			continue
+		}
+		startMatches := findAllIndex(r.start, r.skip, line)
+		endMatches := findAllIndex(r.end, r.skip, line)
+		samePattern := false
+	startLoop:
+		for startIdx := 0; startIdx < len(startMatches); startIdx++ {
+			if startMatches[startIdx][0] < lineLen {
```

Comment:
In what cases is this condition false?

Reply:
To be honest, I can't give a plausible reason why I added this check. :thinking: 
Since a false result would mean the regex parser did something totally wrong it's worth to get informed about it.

Thanks for spotting. 👍